core/primitives/src/receipt.rs

```diff
@@ -158,6 +226,105 @@ impl BorshDeserialize for Receipt {
     }
 }
 
+impl BorshSerialize for StateStoredReceipt {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        // The serialization format for StateStored receipt is as follows:
+        // Byte 1: STATE_STORED_RECEIPT_TAG
+        // Byte 2: STATE_STORED_RECEIPT_TAG
+        // Byte 3: enum version (e.g. V0 => 0_u8)
+        // serialized variant value
+
+        BorshSerialize::serialize(&STATE_STORED_RECEIPT_TAG, writer)?;
+        BorshSerialize::serialize(&STATE_STORED_RECEIPT_TAG, writer)?;
+        match self {
+            StateStoredReceipt::V0(v0) => {
+                BorshSerialize::serialize(&0_u8, writer)?;
+                BorshSerialize::serialize(&v0, writer)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl BorshDeserialize for StateStoredReceipt {
+    fn deserialize_reader<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let u1 = u8::deserialize_reader(reader)?;
+        let u2 = u8::deserialize_reader(reader)?;
+        let u3 = u8::deserialize_reader(reader)?;
+
+        if u1 != STATE_STORED_RECEIPT_TAG || u2 != STATE_STORED_RECEIPT_TAG {
+            let error = Error::new(
+                ErrorKind::Other,
+                "Invalid tag found when deserializing StateStoredReceipt",
+            );
+            return Err(io::Error::new(ErrorKind::InvalidData, error));
```

Comment:
nit: could be nice to add the invalid tag to the error message, just in case. Same with the invalid version.